src/platform/Linux/MdnsImpl.cpp

```diff
@@ -0,0 +1,665 @@
+                                        static_cast<AvahiLookupFlags>(0), HandleBrowse, browseContext.get());
+    if (browser != nullptr)
+    {
+        mBrowseContexts[browseContext.get()] = browseContext;
+    }
+
+    return browser == nullptr ? CHIP_ERROR_INTERNAL : CHIP_NO_ERROR;
+}
+
+MdnsServiceProtocol TruncateProtocolInType(char * type)
+{
+    char * deliminator           = strrchr(type, '.');
+    MdnsServiceProtocol protocol = MdnsServiceProtocol::kMdnsProtocolUnknown;
+
+    if (deliminator != NULL)
+    {
+        if (strcmp("._tcp", deliminator) == 0)
+        {
+            protocol     = MdnsServiceProtocol::kMdnsProtocolTcp;
+            *deliminator = 0;
+        }
+        else if (strcmp("._udp", deliminator) == 0)
+        {
+            protocol     = MdnsServiceProtocol::kMdnsProtocolUdp;
+            *deliminator = 0;
+        }
+    }
+    return protocol;
+}
+
+void MdnsAvahi::HandleBrowse(AvahiServiceBrowser * browser, AvahiIfIndex interface, AvahiProtocol protocol, AvahiBrowserEvent event,
+                             const char * name, const char * type, const char * domain, AvahiLookupResultFlags flags,
+                             void * userdata)
+{
+    (void) flags;
+    BrowseContext * context = static_cast<BrowseContext *>(userdata);
+
+    switch (event)
+    {
+    case AVAHI_BROWSER_FAILURE:
+        context->mCallback(context->mContext, nullptr, 0, CHIP_ERROR_INTERNAL);
+        avahi_service_browser_free(browser);
+        context->mInstance->mBrowseContexts.erase(context);
+        break;
+    case AVAHI_BROWSER_NEW:
+        ChipLogProgress(DeviceLayer, "Avahi browse: cache new");
+        if (strcmp("local", domain) == 0)
+        {
+            MdnsService service;
+
+            strncpy(service.mName, name, sizeof(service.mName));
+            strncpy(service.mType, type, sizeof(service.mType));
+            service.mName[kMdnsNameMaxSize] = 0;
+            service.mType[kMdnsTypeMaxSize] = 0;
+            service.mProtocol               = TruncateProtocolInType(service.mType);
+            context->mServices.push_back(service);
+            break;
+        }
+    case AVAHI_BROWSER_ALL_FOR_NOW:
+        ChipLogProgress(DeviceLayer, "Avahi browse: all for now");
+        context->mCallback(context->mContext, context->mServices.data(), context->mServices.size(), CHIP_NO_ERROR);
+        avahi_service_browser_free(browser);
+        context->mInstance->mBrowseContexts.erase(context);
+        break;
+    case AVAHI_BROWSER_REMOVE:
+        ChipLogProgress(DeviceLayer, "Avahi browse: remove");
+        if (strcmp("local", domain) == 0)
+        {
+            std::remove_if(context->mServices.begin(), context->mServices.end(), [name, type](const MdnsService service) {
+                return name == std::string(service.mName) && type == GetFullType(service.mType, service.mProtocol);
```

Comment:
Use strcmp here instead of allocating a string? We don't have std::string_view until C++17.